uber/models/department.py

```diff
@@ -0,0 +1,271 @@
+from datetime import timedelta
+
+from sideboard.lib import cached_property
+from sideboard.lib.sa import CoerceUTF8 as UnicodeText, \
+    UTCDateTime, UUID
+from sqlalchemy import select
+from sqlalchemy.orm import column_property
+from sqlalchemy.schema import ForeignKey, Table, UniqueConstraint
+from sqlalchemy.types import Boolean, Float, Integer
+
+from uber.config import c
+from uber.models import MagModel
+from uber.models.attendee import Attendee
+from uber.models.types import default_relationship as relationship, \
+    Choice, DefaultColumn as Column
+
+
+__all__ = [
+    'department_membership_job_role', 'department_membership_request',
+    'job_required_job_role', 'Department', 'DepartmentMembership', 'Job',
+    'JobRole', 'Shift']
+
+
+department_membership_job_role = Table(
+    'department_membership_job_role',
+    MagModel.metadata,
+    Column('department_membership_id', UUID,
+           ForeignKey('department_membership.id')),
+    Column('job_role_id', UUID, ForeignKey('job_role.id')))
+
+
+department_membership_request = Table(
+    'department_membership_request',
+    MagModel.metadata,
+    Column('attendee_id', UUID, ForeignKey('attendee.id')),
+    Column('department_id', UUID, ForeignKey('department.id'), nullable=True),
+    UniqueConstraint('attendee_id', 'department_id'))
+
+
+job_required_job_role = Table(
+    'job_required_job_role',
+    MagModel.metadata,
+    Column('job_id', UUID, ForeignKey('job.id')),
+    Column('job_role_id', UUID, ForeignKey('job_role.id')))
+
+
+class DepartmentMembership(MagModel):
+
+    DEPARTMENT_HEAD = 100
```

Comment:
Hmmm... arguably, previous 'statuses' we've used for departments are superceded by the new roles feature. Instead of this column, maybe we just want a way to create a few 'default' roles for each department to ape the same functionality? That way, everything is roles. It's roles all the way down.

Reply:
Yeah, I think "Trusted" should definitely be changed to be a role.

I had also considered making "Department Head" a role, but the department head property is used so pervasively throughout our codebase I thought it would be easier to treat it as a special case. I initially had a dedicated column for department heads:
```python
is_dept_head = Column(Boolean, default=False)
```

I have also seen a user story floating around somewhere to be able to ban people from specific departments without banning them from volunteering entirely. So I considered a dedicated column for banned staffers:
```python
is_banned = Column(Boolean, default=False)
```

But then I had two mutually exclusive boolean values, which is usually a good candidate for combining into a single enum, like the `status` column (which may be better named `type`?).

# User editable roles

I'll update the migration script to create a "Trusted" `JobRole` for each department, and assign the trusted role to the relevant attendees.

The UI will eventually let department heads create and delete their own `JobRoles`. For example, TechOps might create `JobRoles` for Soldering, Audio/Visual, SysAdmin, etc... And department heads will be able to delete the "Trusted" role if they don't think it works for their department.

We _could_ use roles to handle "Department Head", but I am concerned that an admin might decide to delete the "Department Head" role. We could have a flag in the role table like `is_deletable`. Or we could make a special case to prevent the "Department Head" role from being deleted. That seems overly complicated to me though ☹️ 

I ultimately decided that "Department Head" and "Banned" were special enough to merit hardcoding values for them.

So maybe `DepartmentMembership` should look like this:

```python
class DepartmentMembership(MagModel):

    DEPARTMENT_HEAD = 100
    REGULAR = 200
    BANNED = 300

    TYPE_OPTS = [
        (DEPARTMENT_HEAD, 'Department Head'),
        (REGULAR, 'Regular'),
        (BANNED, 'Banned')]
    TYPES = dict(TYPE_OPTS)

    type = Column(Choice(TYPE_OPTS), default=REGULAR)
```

# Separate banned table

Of course, we could have an entirely separate table to track attendees that have been banned from a department. It could track dates, comments, and other things relevant to the attendees banned status. Also, a banned table could be implemented in a separate pull-request to keep the size of this pull-request smallish.

In which case we could make is_dept_head a boolean again:

```python
class DepartmentMembership(MagModel):

    is_dept_head = Column(Boolean, default=False)
```

After typing all this up, I think that might make the most sense!

What do you think?

Reply:
Hmm, that's totally a good point about mutually exclusive roles and delete-ability!

I think the idea of doing a separate banned table and an `is_dept_head` property is great. Quick thing to think about -- do we want to take this opportunity to rename the idea of "dept head" to something possibly more descriptive, like 'has_admin_privileges' or 'is_admin' or 'can_add_roles'? I'm totally good either way, the only reason I suggest it is that it's slightly weird to have a multitude 'department heads' when they're really given the access because they perform specific administrative roles in the department.

Reply:
After sleeping on it for a night, I think I want to keep this pull request as simple as possible. I think what you're describing is a whole new permissions model, which is probably outside the scope of this pull request.

But what you're describing is exactly the direction we eventually want to go! So I think when we do that, we might be able to remove the `is_dept_head` column and replace it with a role 😄

Reply:
Makes sense, and sounds good! :+1: